aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -575,26 +415,38 @@ If a class seems to have too many injected dependencies, this is generally a sig
 
 The container calls <xref:System.IDisposable.Dispose*> for the <xref:System.IDisposable> types it creates. If an instance is added to the container by user code, it isn't disposed automatically.
 
+In the following example, the services are created by the service container and disposed automatically:
+
 ```csharp
-// Services that implement IDisposable:
 public class Service1 : IDisposable {}
 public class Service2 : IDisposable {}
-public class Service3 : IDisposable {}
 
-public interface ISomeService {}
-public class SomeServiceImplementation : ISomeService, IDisposable {}
+public interface IService3 {}
+public class Service3 : IService3, IDisposable {}
 
 public void ConfigureServices(IServiceCollection services)
 {
-    // The container creates the following instances and disposes them automatically:
     services.AddScoped<Service1>();
     services.AddSingleton<Service2>();
-    services.AddSingleton<ISomeService>(sp => new SomeServiceImplementation());
+    services.AddSingleton<IService3>(sp => new Service3());
+}
+```
+
+In the following example:
 
-    // The container doesn't create the following instances, so it doesn't dispose of
-    // the instances automatically:
-    services.AddSingleton<Service3>(new Service3());
-    services.AddSingleton(new Service3());
+* The service instances aren't created by the service container.
+* The intended service lifetimes aren't known by the framework.
+* The framework doesn't dispose of the services automatically.
+* If the services aren't explicitly disposed in developer code, they persist until the app shuts down.
```

Comment:
I think saying “If the services aren't explicitly disposed in developer code, they persist until the app shuts down.” is confusing/misleading. Anything that is not `Dispose`()d prior to shutdown will not somehow automatically be `Dispose()`d during shutdown. Also, `Dispose()`ing something doesn’t cause it to cease to exist (i.e., things persist after they are `Dispose()`d).

For the sake of code correctness and discipline, it would be nice to know how to register a handler for when the container is shut down. If I manually create an instance, then I would like to be able to be informed that the container is being shutdown so I can know the container no longer needs my instance (and, if that was the last thing needing the instance, that would be the time to call `Dispose()` on it).

Reply:
I wouldn't phrase that concern in exactly that way. It doesn't say that the services are disposed at shutdown, it merely says that the services *persist until the app shuts down*, which isn't the same thing. It flows from the prior bullet that the framework does nothing on its own. It could say ...

> If the services aren't explicitly disposed in developer code, they persist until the app's process shuts down.

WRT ...

> how to register a handler for when the container is shut down

See https://github.com/aspnet/AspNetCore.Docs/issues/16773, including *Lock's Gambit*:tm:. On the notion that we'd cover something explicitly here (or cross-link Lock's blog post), note that Nurse or Nelson haven't called for that (*yet*). Let's see what they say on review. I'd like their guidance before making further changes.

Reply:
> I wouldn't phrase that concern in exactly that way. It doesn't say that the services are disposed at shutdown, it merely says that the services _persist until the app shuts down_, which isn't the same thing. It flows from the prior bullet that the framework does nothing on its own. It could say ...
> 
> > If the services aren't explicitly disposed in developer code, they persist until the app's process shuts down.

Adding “process” doesn’t change things. In fact, it makes it less exact—especially if the process lives longer than the server which could be shut down, allowing the instance to be garbage collected (an cease to “persist”—without necessarily ever being `Dispose()`d) prior to the process exiting.

Also, the other thing I don’t like about that is how it read in English. The language of “If x, then y” in English implies “If not x, then not y”. So you’re saying that `Dispose()`ing something causes it to cease to exist (to be fully unreferenced and garbage collected). I’d rather not have the language of the docs suggest that ;-).